packages/medusa/src/services/job-scheduler.ts

```diff
@@ -0,0 +1,124 @@
+import Bull from "bull"
+import Redis from "ioredis"
+import { ConfigModule, Logger } from "../types/global"
+
+type InjectedDependencies = {
+  logger: Logger
+  redisClient: Redis.Redis
+  redisSubscriber: Redis.Redis
+}
+
+type ScheduledJobHandler<T = unknown> = (
+  data: T,
+  eventName: string
+) => Promise<void>
+
+export default class JobSchedulerService {
+  protected readonly config_: ConfigModule
+  protected readonly logger_: Logger
+  protected readonly handlers_: Map<string | symbol, ScheduledJobHandler[]>
+  protected readonly queue_: Bull
+
+  constructor(
+    { logger, redisClient, redisSubscriber }: InjectedDependencies,
+    config: ConfigModule,
+    singleton = true
+  ) {
+    this.config_ = config
+    this.logger_ = logger
+
+    if (singleton) {
+      const opts = {
+        createClient: (type: string): Redis.Redis => {
+          switch (type) {
+            case "client":
+              return redisClient
+            case "subscriber":
+              return redisSubscriber
+            default:
+              if (config.projectConfig.redis_url) {
+                return new Redis(config.projectConfig.redis_url)
+              }
+              return redisClient
+          }
+        },
+      }
+
+      this.handlers_ = new Map()
```

Comment:
suggestion: Since it is only an initialisation, it can be moved to the property declaration on the class